### DISCUSSION TITLE
Install, don't upgrade.

### DISCUSSION THREAD

**Discussion by zmetzing** [Homebrew/discussions#1107] [Everyday usage] [answered]
> zmetzing@cider:~% brew install apcupsd
> Updating Homebrew...
>  ==> Auto-updated Homebrew!
> Updated 3 taps (homebrew/cask-versions, homebrew/core and homebrew/cask).
> ==> New Formulae
> 
>  ( .. 15 minutes later .. )
> 
> ==> Reinstalling python@2 
> ==> Installing dependencies for python@2: sphinx-doc
> ==> Installing python@2 dependency: sphinx-doc
> ==> Pouring sphinx-doc-3.5.3.catalina.bottle.tar.gz
> ==> Caveats
> sphinx-doc is keg-only, which means it was not symlinked into /usr/local,
> because this formula is mainly used internally by other formulae.
> Users are advised to use `pip` to install sphinx-doc.
> 
> If you need to have sphinx-doc first in your PATH, run:
>   echo 'setenv PATH /usr/local/opt/sphinx-doc/bin:$PATH' >> ~/.tcshrc
> 
> ==> Summary
> 🍺  /usr/local/Cellar/sphinx-doc/3.5.3: 3,842 files, 55.4MB
> ==> Installing python@2
> Error: An exception occurred within a child process:
>   FormulaUnavailableError: No available formula with the name "/usr/local/opt/python@2/.brew/python@2.rb".

Thanks. All I really wanted was to get apcupsd installed. Instead, you upgraded (and broke) my world.

*sigh*

--- Zach

**Comment by carlocab** [comment]
`brew upgrade` upgrades all formulae connected to `apcupsd` through the dependency graph. See https://docs.brew.sh/FAQ#why-does-brew-upgrade-formula-also-upgrade-a-bunch-of-other-stuff.

To avoid this happening again, you can try extracting `python@2` to your own tap with `brew extract python@2`.

**Comment by tlk** [comment, marked as answer]
You may find HOMEBREW_NO_AUTO_UPDATE  useful, see https://github.com/Homebrew/discussions/discussions/578#discussioncomment-295595

The advantage of being able to install a piece of software without fetching all the latest updates (and potentially upgrading homebrew itself) is that your system may do just what you was expecting: install that one piece of software. The disadvantage, as I understand it, is that many users used to forget to run `brew update` manually, and this was the primary reason for the homebrew maintainers to make the change to auto-update-on-install.

**Reply by SMillerDev** [reply]
The downside is that it often breaks people's system because they have incompatible versions of software installed. I'd really only ever recommend setting this in a CI system where you update regularly, otherwise you're likely to run into hard to debug issues.

**Reply by tlk** [reply]
Ok, I don't understand - how will the issue of incompatible versions appear using HOMEBREW_NO_AUTO_UPDATE?

When I run `brew update` manually, it is my understanding that homebrew pulls down the most recent formulas. Won't all these formulas be compatible internally? If I come back a month later and run `brew install foo` the foo-package may depend on x, y and z but assuming the formulas I pulled down a month ago were functional, the foo package should install fine after downloading and installing (possibly outdated, but compatible) versions of x, y and z. I do understand that some source/binary packages may no longer be available from various CDNs etc, but if it is still available - where would the incompatibility arise?

Worst case I get an error telling me that file x.tgz is no longer available, and I can then decide to run `brew update` knowing that this may upgrade (or break) the entire world and its dog.

I guess I am missing something?

**Reply by zmetzing** [reply]
> The downside is that it often breaks people's system because they have incompatible versions of software installed. I'd really only ever recommend setting this in a CI system where you update regularly, otherwise you're likely to run into hard to debug issues.

This is false logic. If I've "upgrad"ed in the recent past, the formulae on my system should be internally consistent. Don't force upgrades on me; If something would break horribly on my system, warn me and bail out.

FreeBSD ports works this way. I get a tree of ports and I can then build from that tree as long as I want (assuming source files are still fetchable from the 'net).

**Reply by SMillerDev** [reply]
> Ok, I don't understand - how will the issue of incompatible versions appear using HOMEBREW_NO_AUTO_UPDATE?

Because people don't update manually at the exact right time and homebrew won't either.

> Won't all these formulas be compatible internally?

The formula files, yes. The software might not. Since homebrew is an evergreen package manager we only ever build against the latest version. If you install python 3.9.1 and 3.9.2 is released that changes something SQLite uses for example, your SQLite will not work after install (because it is build against 3.9.2)

> This is false logic. If I've "upgrad"ed in the recent past, the formulae on my system should be internally consistent. Don't force upgrades on me; If something would break horribly on my system, warn me and bail out.

No, that's not the user experience Homebrew wants for its users. Maybe macports is closer to how you think package managers should work?

**Reply by zmetzing** [reply]
> No, that's not the user experience Homebrew wants for its users. Maybe macports is closer to how you think package managers should work?

Ah, and there's the rub: You've forgotten that the _users_ are the people that _want_ it to work this way, not the other way around. You're not here to foist a solution on your customers, you're here to delight them. Unfortunately, since there is no absolute monetary feedback for open source, you have no pressure to do that.

**Reply by tlk** [reply]
> The formula files, yes. The software might not. Since homebrew is an evergreen package manager we only ever build against the latest version. If you install python 3.9.1 and 3.9.2 is released that changes something SQLite uses for example, your SQLite will not work after install (because it is build against 3.9.2)

Ok, but having - say - month-old formula files on my system, I would be surprised if those month-old formulas would a) know about a build from today and b) be allowed to install it. Would the checksums not prevent this from happening?

I mean, my mental model might be off here, but I really don't understand the argument.

**Reply by SMillerDev** [reply]
> You've forgotten that the users are the people that want it to work this way, not the other way around.

The amount of complaints/issues we had about outdated dependencies in the old flow (that you are suggesting) far outweighs the ~2 people a year that prefers to not have their system auto-update.

I get that not everyone wants their package manager to do the heavy lifting for them, but the sad reality is that we can't please everyone.

**Reply by iMichka** [reply]
> customers /  you have no pressure to do that.

That's the point. You are not paying anything. And 99% of our users are happy. So for the 1% remaining, feel free to use any other tool that better suits your need.

**Reply by zmetzing** [reply]
> > customers /  you have no pressure to do that.
> 
> That's the point. You are not paying anything. And 99% of our users are happy. So for the 1% remaining, feel free to use any other tool that better suits your need.

Thanks for making HomeBrew's position clear. Good luck!

**Reply by SMillerDev** [reply]
> a) know about a build from today

If you run `brew update` in between it will (without upgrading)

> b) be allowed to install it

We don't checksum dependencies. We just specify `depends_on "dependency"` and it's up to brew to resolve that to the latest version.

**Reply by tlk** [reply]
> > a) know about a build from today
> 
> If you run `brew update` in between it will (without upgrading)

Yes and that is exactly the point. With HOMEBREW_NO_AUTO_UPDATE I can choose not to run `brew update`. And because of that, I can run `brew install foo` without any inconsistencies.

Right?

**Reply by SMillerDev** [reply]
> With HOMEBREW_NO_AUTO_UPDATE I can choose not to run brew update. And because of that, I can run brew install foo without any inconsistencies.

Only if you never update your OS either and barring any race conditions with PRs being merged sequentially instead of synchronous, yes.

**Reply by tlk** [reply]
> The amount of complaints/issues we had about outdated dependencies in the old flow (that you are suggesting) far outweighs the ~2 people a year that prefers to not have their system auto-update.

I don't want to get into a heated argument and I do not dispute your experience about the old default, but to me #1112 looks related so that is more like 2 people within 1 day.

**Reply by tlk** [reply]
> Only if you never update your OS either and barring any race conditions with PRs being merged sequentially instead of synchronous, yes.

Ok, thanks for explaining.

I don't understand how merging PR's are related to this?

Are you thinking about the situation where my month-old `brew update` took place in the middle of some complicated merge process so the formulas were not compatible at that point (but got fixed say 10 minutes later)?

Just to clarify, I am not arguing that installing month-old software is desirable, it's just that I don't see how using HOMEBREW_NO_AUTO_UPDATE is so awful. Afterall, if `brew install foo` does not work for some reason, I can always run `brew update` followed by `brew install foo`.

**Reply by SMillerDev** [reply]
> Are you thinking about the situation where my month-old brew update took place in the middle of some complicated merge process so the formulas were not compatible at that point (but got fixed say 10 minutes later)?

Exactly this

> Afterall, if brew install foo does not work for some reason, I can always run brew update followed by brew install foo.

The chance that brew install fails is pretty slim. It's much more likely that some specific part of a software starts misbehaving slightly. Which leads to people reporting issues, which takes up the scarce maintainer time.

**Reply by tlk** [reply]
Okay, I understand. Thanks for the clarification.

**Reply by elder-n00b** [reply]
I've run into this problem too.

I also think a simple "dear user, did you remember to update or else do not complain" banner on each run could have worked to lower the pressure on maintainers, the same way I get to see endless times that my Xcode and macOS are old and unsupported.
 
The thing is, people still use older software and OS versions, if anything because they cannot / do not want to replace their hardware (plus other reasons), both at home and in professional settings, and in those circumstances a `brew install` often either fails when it could have worked or worse proceeds when it shouldn't have, damaging a working configuration in the process, and there's no obvious way back, because the same "thou shalt upgrade" policy -- though with hindsight I guess one could use `git` to check out an older formulae repo revision, is this correct?

As I understand (now, after having banged my head against the wall) a similar "dear user, be careful and remember there are options to not auto update/upgrade" would have sufficed too. 

I'm often able to get around upgrade issues with `HOMEBREW_NO_AUTO_UPDATE`, `brew pin <formula>`, and/or `--ignore-dependencies`, though I think (not tried yet) the proper way would be to locally fork the formulae repository and maintain it. I see there's `brew extract` too, I'll have to read up on that. Yeah, RTFM is always the right thing to do, but I wish this "upgrade hell" related info was a bit more prominent (with the obvious "you're on your own and do not bother us" disclaimer). BTW when is it advisable to do a `brew edit` and when a `brew extract`? 

After all, one uses a package manager not just to avoid the hassle of downloading files and typing long command lines, but because he trusts the package manager to do the right thing and not wreck anything by mistake. 

I'll trust you with the "99% of our users are happy" (all in all, I too am happy with brew, and I enjoy a free beer :-) ), but I propose these:
1. better warnings and more prominent documentation about this potential issue.
2. optional dry-run when there are dependencies to upgrade, followed by "proceed? Y/N". <br> optional because interactivity is not always desirable, in particular when not predictable.
3. a little more information from `brew deps --tree`: version available, version installed, is it pinned, √/X mark like in `brew info` (it would all fit in one line), maybe only when doing `brew deps --tree -v` (as of now, the verbose flag doesn't seem to do anything in this case).
4. a lock file somewhere to avoid the inconsistency issue described above as: 
> Are you thinking about the situation where my month-old brew update took place in the middle of some complicated merge process so the formulas were not compatible at that point (but got fixed say 10 minutes later)?
5. a documented, non hacky way to mark software installed by other means as a valid (and pinned) dependency (at user's total responsibility of course). I see there's a hint to install such software in the cellar and `brew link`, but doesn't seem always applicable (another package manager, some binary distributions, some build tools eg. fpcupdeluxe for fpc, I'm wary they won't play nice with each other).
6. maybe semi-official formulae repositories for older systems (homebrew would still maintain only the "most recent" one, users would contribute to the "legacy" ones, initially forked from earlier revisions of the main one, again all at the user's risk and responsibility). This one very maybe, it could potentially cause more trouble.

Maybe I can help (I know a little Ruby and none of my proposals seem too hard to implement in theory), maybe not ("works for me" does not qualify as testing), in any case I'd have some homework to do before I can attempt. I'll get in touch in case I manage to have some working code to contribute. 
By the way brew is useful to me also because I have an old system for which building from sources is often a necessity and brew helps me with that very much, so this problem affects me and it's likely I'll give it a shot. It would seem I'm not alone (at least #578 and #1112 are related). 
I believe such small changes would increase the usefulness in corner cases without hampering the "evergreen" philosophy.

A note about python@3.9 which is often a source of trouble: brew installed 3.9.1_2 (which I subsequently pinned) but since a while fails to install/upgrade formulae depending on it without the `--ignore-dependencies` workaround. Many seem to be happy with any 3.9 being available (of course I'm missing out on Python fixes and updates), at least the ones which are Python source code and do not link anything.
Would a more lenient dependency version check make sense in those formulae?

A final (and somewhat off-topic) remark: a few casks are... well... do they at least support development economically? 
And some got installed/upgraded even if incompatible with the OS version (formula problem I guess, perhaps already fixed) though I forgot which ones. At least the "drag this icon" method gives me a chance to see the "no way" road sign in those cases. Maybe there's a way to ask the OS if an app is compatible before moving it in place?

**Reply by SMillerDev** [reply]
> I also think a simple "dear user, did you remember to update or else do not complain" banner on each run could have worked to lower the pressure on maintainers, the same way I get to see endless times that my Xcode and macOS are old and unsupported.

It doesn't because people don't read. You might, but there are still countless people that file issues when we tell them not to.

> though with hindsight I guess one could use git to check out an older formulae repo revision, is this correct?

The best bet would be to checkout both an older brew tool and formulae version since those can be tightly linked. But an even better solution would be `brew extract`.

> BTW when is it advisable to do a brew edit and when a brew extract?

Brew edit is really only recommended if you want to upstream your changes after.

> better warnings and more prominent documentation about this potential issue.

Pull requests are welcome although I don't see how we can better warn that your install process will install things.

> optional dry-run when there are dependencies to upgrade, followed by "proceed? Y/N".
optional because interactivity is not always desirable, in particular when not predictable.

Brew has never done interactive steps and that is highly unlikely to change. But you can do `brew upgrade --dry-run` if you want to see the result of an upgrade.

> a lock file somewhere to avoid the inconsistency issue described above as:

and 
> a documented, non hacky way to mark software installed by other means as a valid (and pinned) dependency

That conflicts with brew being an evergreen package manager so that won't happen.

> maybe semi-official formulae repositories for older systems

You can start a tap and make a PR to add it to https://docs.brew.sh/Interesting-Taps-and-Forks

**Reply by elder-n00b** [reply]
> there are still countless people that file issues when we tell them not to.

I understand. Sadly there's no easy way to stop users from making mistakes and then complaining to someone else for providing the tools. I feel a little guilty here... I run into a problem because a) I did not read the documentation thoroughly and b) expected nothing can break when building from sources anyways, only a failed build and goodbye (I still find it a reasonable expectation).

> I don't see how we can better warn that your install process will install things.

It's quite obvious that install means install, I mean warn that an implicit upgrade may not be safe. Well, I know that by now.

> That conflicts with brew being an evergreen package manager so that won't happen.

I don't see how point 4, the one about the lock file, would conflict in any way. It would only prevent "updating at the wrong time" and delay the update to the next invocation. I think in that scenario described the local repo would be inconsistent, evergreen or not, at least for that invocation which may well happen to be an install or upgrade. But if unlikely enough, and simply repeating the operation would fix anything that might have broke, indeed probably it's not worth it.

Re point 5, marking stuff as dependencies, my bad. Extract, requirements and resources are what I was after.

I'll go the tap route as advised and share it if it works well. It goes without saying, it would have a prominent disclaimer that it's borderline abuse and totally unsupported by the homebrew devs.

Thanks and best regards.
